kuma/wiki/tests/test_views_document.py

```diff
@@ -590,3 +635,177 @@ def test_tags_not_show_while_empty(root_doc, client, wiki_user):
+    assert 'public' in response['Cache-Control']
+    assert 's-maxage' in response['Cache-Control']
+    assert response['Location'].endswith(trans_doc.get_absolute_url() + params)
+
+
+def test_redirect_with_no_slug(db, client):
+    """Bug 775241: Fix exception in redirect for URL with ui-locale"""
+    url = '/en-US/docs/en-US/'
+    response = client.get(url)
+    assert response.status_code == 404
+    assert 'max-age=0' in response['Cache-Control']
+    assert 'no-cache' in response['Cache-Control']
+    assert 'no-store' in response['Cache-Control']
+    assert 'must-revalidate' in response['Cache-Control']
+
+
+@pytest.mark.parametrize(
+    'http_method', ['get', 'put', 'delete', 'options', 'head'])
+@pytest.mark.parametrize(
+    'endpoint', ['wiki.subscribe', 'wiki.subscribe_to_tree'])
+def test_watch_405(client, root_doc, endpoint, http_method):
+    """Watch document with HTTP non-POST request results in 405."""
+    url = reverse(endpoint, locale=root_doc.locale, args=[root_doc.slug])
+    response = getattr(client, http_method)(url)
+    assert response.status_code == 405
+    assert 'max-age=0' in response['Cache-Control']
+    assert 'no-cache' in response['Cache-Control']
+    assert 'no-store' in response['Cache-Control']
+    assert 'must-revalidate' in response['Cache-Control']
+
+
+@pytest.mark.parametrize(
+    'endpoint', ['wiki.subscribe', 'wiki.subscribe_to_tree'])
+def test_watch_login_required(client, root_doc, endpoint):
+    """Watch document with HTTP non-POST request results in 405."""
+    url = reverse(endpoint, locale=root_doc.locale, args=[root_doc.slug])
+    response = client.post(url)
+    assert response.status_code == 302
+    assert 'max-age=0' in response['Cache-Control']
+    assert 'no-cache' in response['Cache-Control']
+    assert 'no-store' in response['Cache-Control']
+    assert 'must-revalidate' in response['Cache-Control']
+    assert response['Location'].endswith(
+        reverse('account_login') + '?next=' + quote(url))
+
+
+@pytest.mark.parametrize(
+    'endpoint,event', [('wiki.subscribe', EditDocumentEvent),
+                       ('wiki.subscribe_to_tree', EditDocumentInTreeEvent)],
+    ids=['subscribe', 'subscribe_to_tree'])
+def test_watch_unwatch(user_client, wiki_user, root_doc, endpoint, event):
+    """Watch and unwatch a document."""
+    url = reverse(endpoint, locale=root_doc.locale, args=[root_doc.slug])
+    # Subscribe
+    response = user_client.post(url)
+    assert response.status_code == 302
+    assert 'max-age=0' in response['Cache-Control']
+    assert 'no-cache' in response['Cache-Control']
+    assert 'no-store' in response['Cache-Control']
+    assert 'must-revalidate' in response['Cache-Control']
+    assert response['Location'].endswith(
+        reverse('wiki.document', locale=root_doc.locale, args=[root_doc.slug]))
+    assert event.is_notifying(wiki_user, root_doc), 'Watch was not created'
+
+    # Unsubscribe
+    response = user_client.post(url)
```

Comment:
😢  it's the same endpoint to subscribe and unsubscribe 🤦‍♂️

Reply:
I couldn't believe when I first saw the tests. 🤦‍♂️